quic/s2n-quic-transport/src/stream/controller.rs

```diff
@@ -234,85 +298,20 @@ impl transmission::interest::Provider for Controller {
         &self,
         query: &mut Q,
     ) -> transmission::interest::Result {
-        self.bidi_controller.transmission_interest(query)?;
-        self.uni_controller.transmission_interest(query)?;
-        Ok(())
-    }
-}
-
-/// The controller pair consists of both local_initiated and remote_initiated
-/// controllers that are both notified when a stream is opened, regardless
-/// of which side initiated the stream.
-#[derive(Debug)]
-struct ControllerPair {
-    stream_id: StreamId,
-    local_initiated: LocalInitiated,
-    remote_initiated: RemoteInitiated,
-}
-
-impl ControllerPair {
-    #[inline]
-    fn on_open_stream(&mut self) {
-        self.local_initiated.on_open_stream();
-        self.remote_initiated.on_open_stream();
-    }
-
-    #[inline]
-    fn on_close_stream(&mut self) {
-        self.local_initiated.on_close_stream();
-        self.remote_initiated.on_close_stream();
-    }
-
-    #[inline]
-    pub fn on_packet_ack<A: ack::Set>(&mut self, ack_set: &A) {
-        self.remote_initiated.on_packet_ack(ack_set);
-        self.local_initiated.on_packet_ack(ack_set);
-    }
-
-    #[inline]
-    pub fn on_packet_loss<A: ack::Set>(&mut self, ack_set: &A) {
-        self.remote_initiated.on_packet_loss(ack_set);
-        self.local_initiated.on_packet_loss(ack_set);
-    }
-
-    #[inline]
-    pub fn on_timeout(&mut self, now: Timestamp) {
-        self.local_initiated.on_timeout(now);
-    }
-
-    #[inline]
-    pub fn on_transmit<W: WriteContext>(&mut self, context: &mut W) -> Result<(), OnTransmitError> {
-        // Only the stream_type from the StreamId is transmitted
-        self.remote_initiated.on_transmit(self.stream_id, context)?;
-        self.local_initiated.on_transmit(self.stream_id, context)
-    }
-
-    /// This method is called when the stream manager is closed. All wakers will be woken
-    /// to unblock waiting tasks.
-    pub fn close(&mut self) {
-        self.local_initiated.close();
-        self.remote_initiated.close();
-    }
-}
-
-impl timer::Provider for ControllerPair {
-    #[inline]
-    fn timers<Q: timer::Query>(&self, query: &mut Q) -> timer::Result {
-        self.local_initiated.timers(query)?;
+        self.local_bidi_controller.transmission_interest(query)?;
+        self.remote_bidi_controller.transmission_interest(query)?;
+        self.local_uni_controller.transmission_interest(query)?;
+        self.remote_uni_controller.transmission_interest(query)?;
         Ok(())
     }
 }
 
-impl transmission::interest::Provider for ControllerPair {
-    #[inline]
-    fn transmission_interest<Q: transmission::interest::Query>(
-        &self,
-        query: &mut Q,
-    ) -> transmission::interest::Result {
-        self.remote_initiated.transmission_interest(query)?;
-        self.local_initiated.transmission_interest(query)?;
-        Ok(())
-    }
+#[derive(Debug, Copy, Clone)]
+enum StreamDirection {
```

Comment:
The enum you replaced this with had some comments describing each type of stream direction, would be good to have comments for these as well